File: ai-interview-service/src/services/interview.service.js
import InterviewQuestion from "../models/InterviewQuestion.js";
import InterviewReport from "../models/InterviewReport.js";
import InterviewSession from "../models/InterviewSession.js";

// ─── Session Operations ─────────────────────────────────────

export const createSession = async ({ email, company, role, resumeText, resumeData }) => {
  return InterviewSession.create({
    email,
    company,
    role,
    resumeText,
    resumeSkills: resumeData.technical_skills || [],
    resumeExperience: resumeData.experience_summary || "",
  });
};

export const getSessionById = async (sessionId) => {
  return InterviewSession.findById(sessionId);
};

export const updateSessionSections = async (sessionId, sections, totalQuestions) => {
  return InterviewSession.findByIdAndUpdate(
    sessionId,
    {
      sections: sections.map((s) => ({
        title: s.title,
        description: s.description || "",
        questionCount: (s.questions || []).length,
      })),
      totalQuestions,
    },
    { new: true }
  );
};

// ─── Question Operations ────────────────────────────────────

export const addQuestion = async ({ sessionId, questionNumber, questionText, sectionIndex, sectionTitle, difficulty }) => {
  return InterviewQuestion.create({
    sessionId,
    questionNumber,
    questionText,
    sectionIndex: sectionIndex || 0,
    sectionTitle: sectionTitle || "",
    difficulty: difficulty || "medium",
  });
};

export const addAllQuestions = async (sessionId, sections) => {
  const questions = [];
  let questionNumber = 0;

  for (let sectionIndex = 0; sectionIndex < sections.length; sectionIndex++) {
    const section = sections[sectionIndex];
    for (const q of section.questions || []) {
      questions.push({
        sessionId,
        questionNumber,
        questionText: q.question,
        sectionIndex,
        sectionTitle: section.title,
        difficulty: q.difficulty || "medium",
        type: q.type || "text",
        coding: q.coding || undefined,
      });
      questionNumber++;
    }
  }

  return InterviewQuestion.insertMany(questions);
};

export const getQuestions = async (sessionId) => {
  return InterviewQuestion.find({ sessionId }).sort({ questionNumber: 1 });
};

export const getQuestionByNumber = async (sessionId, questionNumber) => {
  return InterviewQuestion.findOne({ sessionId, questionNumber });
};

export const getCurrentQuestionIndex = async (sessionId) => {
  const session = await getSessionById(sessionId);
  return session ? session.currentQuestionIndex : 0;
};

export const updateQuestionIndex = async (sessionId, nextIndex) => {
  // Also update section index
  const questions = await getQuestions(sessionId);
  const currentQ = questions[nextIndex];
  const sectionIndex = currentQ ? currentQ.sectionIndex : 0;

  return InterviewSession.findByIdAndUpdate(
    sessionId,
    { currentQuestionIndex: nextIndex, currentSectionIndex: sectionIndex },
    { new: true }
  );
};

export const submitAnswer = async ({ sessionId, questionNumber, answer, feedback, score, strengths, improvements, modelAnswer }) => {
  return InterviewQuestion.findOneAndUpdate(
    { sessionId, questionNumber },
    {
      userAnswer: answer,
      feedback,
      score,
      strengths: strengths || [],
      improvements: improvements || [],
      modelAnswer: modelAnswer || "",
      answeredAt: new Date(),
    },
    { new: true }
  );
};

// ─── Counter Question Operations ────────────────────────────

export const addCounterQuestion = async (sessionId, questionNumber, counterQuestion) => {
  return InterviewQuestion.findOneAndUpdate(
    { sessionId, questionNumber },
    {
      $push: {
        counterQuestions: {
          question: counterQuestion,
          askedAt: new Date(),
        },
      },
    },
    { new: true }
  );
};

export const submitCounterAnswer = async (sessionId, questionNumber, counterIndex, answer, feedback) => {
  const question = await getQuestionByNumber(sessionId, questionNumber);
  if (!question || !question.counterQuestions[counterIndex]) return null;

  question.counterQuestions[counterIndex].userAnswer = answer;
  question.counterQuestions[counterIndex].feedback = feedback;
  await question.save();
  return question;
};

// ─── Conversation History ───────────────────────────────────

export const addToConversationHistory = async (sessionId, entry) => {
  return InterviewSession.findByIdAndUpdate(
    sessionId,
    {
      $push: {
        conversationHistory: {
          role: entry.role,
          content: entry.content,
          type: entry.type || "question",
          timestamp: new Date(),
        },
      },
    },
    { new: true }
  );
};

export const getConversationHistory = async (sessionId) => {
  const session = await getSessionById(sessionId);
  return session?.conversationHistory || [];
};

// ─── Report Operations ──────────────────────────────────────

export const getReportBySessionId = async (sessionId) => {
  return InterviewReport.findOne({ sessionId });
};

export const createReport = async ({
  sessionId,
  overallScore,
  summary,
  strengths,
  weaknesses,
  recommendations,
  sectionScores,
  skillAssessment,
}) => {
  const report = await InterviewReport.create({
    sessionId,
    overallScore,
    summary,
    strengths,
    weaknesses,
    recommendations,
    sectionScores: sectionScores || [],
    skillAssessment: skillAssessment || [],
  });

  await InterviewSession.findByIdAndUpdate(sessionId, {
    status: "completed",
    completedAt: new Date(),
  });

  return report;
};

export const endInterview = async (sessionId) => {
  return InterviewSession.findByIdAndUpdate(
    sessionId,
    { status: "completed", completedAt: new Date() },
    { new: true }
  );
};

File: ai-interview-service/app/services/interview_service.py
import uuid
from datetime import datetime
from typing import Optional, List
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from app.models.models import InterviewSession, InterviewQuestion, InterviewReport


class InterviewService:
    """Service for managing interview sessions"""

    @staticmethod
    async def create_session(db: AsyncSession, email: str, company: str, role: str,
                             resume_text: str, resume_data: dict) -> InterviewSession:
        """Create a new interview session"""
        session_id = str(uuid.uuid4())

        session = InterviewSession(
            id=session_id,
            email=email,
            company=company,
            role=role,
            resume_text=resume_text,
            resume_skills=resume_data.get("technical_skills", []),
            resume_experience=resume_data.get("experience_summary", ""),
            status="active"
        )
        db.add(session)
        await db.commit()
        await db.refresh(session)
        return session

    @staticmethod
    async def get_session(db: AsyncSession, session_id: str) -> Optional[InterviewSession]:
        """Retrieve an interview session"""
        result = await db.execute(
            select(InterviewSession).filter(InterviewSession.id == session_id)
        )
        return result.scalars().first()

    @staticmethod
    async def add_question(db: AsyncSession, session_id: str, question_number: int,
                           question_text: str) -> InterviewQuestion:
        """Add a question to the interview"""
        question_id = str(uuid.uuid4())

        question = InterviewQuestion(
            id=question_id,
            session_id=session_id,
            question_number=question_number,
            question_text=question_text
        )
        db.add(question)
        await db.commit()
        await db.refresh(question)
        return question

    @staticmethod
    async def submit_answer(db: AsyncSession, session_id: str, question_number: int,
                            answer: str, feedback: str, score: int):
        """Store answer and evaluation for a question"""
        result = await db.execute(
            select(InterviewQuestion).filter(
                InterviewQuestion.session_id == session_id,
                InterviewQuestion.question_number == question_number
            )
        )
        question = result.scalars().first()

        if question:
            question.user_answer = answer
            question.feedback = feedback
            question.score = score
            await db.commit()
            await db.refresh(question)

        return question

    @staticmethod
    async def get_current_question_index(db: AsyncSession, session_id: str) -> int:
        """Get the current question index for a session"""
        session = await InterviewService.get_session(db, session_id)
        if session:
            return session.current_question_index
        return 0

    @staticmethod
    async def update_question_index(db: AsyncSession, session_id: str, index: int):
        """Update the current question index"""
        session = await InterviewService.get_session(db, session_id)
        if session:
            session.current_question_index = index
            await db.commit()

    @staticmethod
    async def get_all_questions(db: AsyncSession, session_id: str) -> List[InterviewQuestion]:
        """Get all questions for a session"""
        result = await db.execute(
            select(InterviewQuestion).filter(
                InterviewQuestion.session_id == session_id
            ).order_by(InterviewQuestion.question_number)
        )
        return result.scalars().all()

    @staticmethod
    async def create_report(db: AsyncSession, session_id: str, overall_score: int,
                            summary: str, strengths: list, weaknesses: list,
                            recommendations: list) -> InterviewReport:
        """Create an interview report"""
        report_id = str(uuid.uuid4())

        report = InterviewReport(
            id=report_id,
            session_id=session_id,
            overall_score=overall_score,
            summary=summary,
            strengths=strengths,
            weaknesses=weaknesses,
            recommendations=recommendations
        )
        db.add(report)

        session = await InterviewService.get_session(db, session_id)
        if session:
            session.status = "completed"
            session.completed_at = datetime.utcnow()

        await db.commit()
        await db.refresh(report)
        return report

    @staticmethod
    async def get_report(db: AsyncSession, session_id: str) -> Optional[InterviewReport]:
        """Retrieve an interview report"""
        result = await db.execute(
            select(InterviewReport).filter(InterviewReport.session_id == session_id)
        )
        return result.scalars().first()
